docs: add ZK chain tutorial (#36)
zkSync-Community-Hub/community-code: content/tutorials/custom-zk-chain/20.customizing-your-chain.md

---
title: Customizing your ZK chain
description: Create a ZK chain with a custom ERC20 base token.

---

One way you can customize your ZK chain is by changing the base token used for gas from ETH to an ERC20 token.
Let's try customizing a new chain by using a custom ERC20 base token.

For the first release of the ZK stack, the base tokens you can use for your chain must be whitelisted.
The whitelisted addresses are stored and checked in the `BridgeHub` contract on the L1.

> In the future it will be possible to add a new token in a permissionless proccess.

Reviewer: This is the first mention of ZK stack in this ZK chain guide. The wording felt a little confusing as it sounded like I (as the reader going through the tutorial) am releasing a zk stack and have to whitelist. 

Do we mean more along the lines of:
```suggestion
With the initial release of ZK stack, base tokens must be whitelisted for use on chains.
```

> In the future it will be possible to add a new token in a permissionless proccess.

Reviewer: ```suggestion
In the future it will be possible to add a new token in a permissionless process.
```

```

Select the option **Create a Typescript project** and accept the default options for everything else.

Run the command below to install the necessary dependencies:

::code-group

```bash [npm]
npm i -D typescript ts-node @openzeppelin/contracts @nomicfoundation/hardhat-ethers @typechain/ethers-v6 @typechain/hardhat typechain ethers dotenv
```

```bash [yarn]
yarn add -D typescript ts-node @openzeppelin/contracts @nomicfoundation/hardhat-ethers @typechain/ethers-v6 @typechain/hardhat typechain ethers dotenv
```

::

Once installed, replace your existing config in `hardhat.config.ts` with the config below:

```ts
import { HardhatUserConfig } from "hardhat/config";
import "@nomicfoundation/hardhat-toolbox";

import dotenv from "dotenv";
dotenv.config();

const config: HardhatUserConfig = {
  solidity: "0.8.20",
  networks: {
        localRethNode: {
          url: "http://localhost:8545",
          accounts: [process.env.WALLET_PRIVATE_KEY as any],
        },
      },
};

export default config;
```

Next, create a `.env` file with:

```bash
touch .env
```

and add the `WALLET_PRIVATE_KEY` environment variable with the private key of the rich wallet we've been using:

```env

> WALLET_PRIVATE_KEY=0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110

Reviewer: I now realize this is just a representation of the address of whose private key we want to use, but it feels a little confusing seeing it this way. It's very tempting to just copy/paste and I see a lot of users doing that without paying attention to what is being represented here.

Author: I added a couple notes for this, but not sure it addresses your concern

    constructor(string memory name, string memory symbol) ERC20(name, symbol) Ownable(msg.sender) {
        _mint(msg.sender, 100 * 10 ** decimals());
    }

    function mint(address to, uint256 amount) public onlyOwner {
        _mint(to, amount);
    }
}
```

Next, let's update the ignition module.
Rename the module file with the command below:

```bash
mv ignition/modules/Lock.ts ignition/modules/CustomBaseToken.ts
```

Then, replace the module file with the code below:

```ts
import { buildModule } from "@nomicfoundation/hardhat-ignition/modules";

const CustomBaseTokenModule = buildModule("CustomBaseTokenModule", (m) => {
  const tokenName = m.getParameter("name", "ZK Base Token");
  const tokenSymbol = m.getParameter("symbol", "ZKBT");

  const baseToken = m.contract("CustomBaseToken", [tokenName, tokenSymbol]);

  return { baseToken };
});

export default CustomBaseTokenModule;
```

To deploy the token contract, run:

```bash
npx hardhat ignition deploy ./ignition/modules/CustomBaseToken.ts --network localRethNode
```

Select `y` to confirm to deploy the contract to the `localRethNode`.
After deploying, the token contract address should be logged in your console.

> Let's verify that the tokens were minted to the deployer address using the `cast` CLI:

Reviewer: ```suggestion
Let's verify that the tokens were minted to the deployer address using the Foundry `cast` CLI:
```

```bash

> cast balance --erc20 0xYOUR_TOKEN_ADDRESS 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049 --rpc-url http://localhost:8545

Reviewer: ```suggestion
cast balance --erc20 <0xYOUR_TOKEN_ADDRESS> 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049 \
--rpc-url http://localhost:8545
```

```

Now that you have some tokens, the next step is to send some to the governor's address on the L1 so that you can register and deploy your chain.
You can find the address for the governor in `my_elastic_chain/configs/wallets.yaml` under `governor`.

Run the command below to use `cast` to transfer some of your ERC20 tokens to the governor's address on the L1.

```bash

> cast send 0xYOUR_TOKEN_ADDRESS "transfer(address,uint256)" 0x_YOUR_GOVERNOR_ADDRESS 5000000000000000000 --private-key 0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110 --rpc-url http://localhost:8545 --gas-price 30000000000

Reviewer: ```suggestion
cast send <0xYOUR_TOKEN_ADDRESS> "transfer(address,uint256)" \ 
<0x_YOUR_GOVERNOR_ADDRESS> 5000000000000000000 \
--private-key <private-key> \
--rpc-url http://localhost:8545 \
--gas-price 30000000000
```

```

To verify that this worked, check the balance of the governor's address:

```bash

> cast balance --erc20 0xYOUR_TOKEN_ADDRESS 0x_YOUR_GOVERNOR_ADDRESS --rpc-url http://localhost:8545

Reviewer: ```suggestion
cast balance --erc20 <0xYOUR_TOKEN_ADDRESS> \
<0x_YOUR_GOVERNOR_ADDRESS> \ 
--rpc-url http://localhost:8545
```

```

## Creating a new chain

Now that your ERC20 token is deployed and the governor's address is funded with that token, you can create a new chain.

First, shut down the node server running for `zk_chain_1`.

Move back into the `my_elastic_chain` ecosystem folder and run the `chain create` command:

```bash
zk_inception chain create
```

This time, use the answers below:

```bash
$ zk_inception chain create

┌   ZKsync toolbox
│
◇  What do you want to name the chain?
│  custom_zk_chain
│
◇  What's the chain id?
│  272
│
◇  Select how do you want to create the wallet
│  Localhost
│
◇  Select the prover mode
│  NoProofs
│
◇  Select the commit data generator mode
│  Rollup
│
◇  Select the base token to use
│  Custom
│
◇  What is the token address?

> │  0x_YOUR_TOKEN_ADDRESS

Reviewer: ```suggestion
│  <0x_YOUR_TOKEN_ADDRESS>
```

│
◇  What is the base token price nominator?
│  1
│
◇  What is the base token price denominator?
│  1
│
◇  Set this chain as default?
│  Yes
```

The base token nominator and denominator are used to define the relation of the base token price with ETH.
For example, if we set the nominator to 20 and the denominator to 1, together the relation is 20/1,
this would mean that 20 tokens would be given the equivalent value as 1 ETH for gas.
For testing purposes, we'll just use a 1:1 ratio.

Next, initialize the chain in the ecosystem with:

```bash
zk_inception chain init
```

> This time select no for the option to deploy a paymaster contract.

Reviewer: ```suggestion
This time select *no* for the option to deploy a paymaster contract.
```

Use the default answers for all other prompts.

Now that the chain is initialized, you can start the chain server:

```bash
zk_inception server
```

## Bridging the base token to your chain

The last step is to bridge some ERC20 tokens from the L1 to the new chain.
Base tokens can not be minted on the L2 without being backed by the corresponding L1 amount. In the future, there will be more flexibility for this.

Open a new terminal and use the `zksync-cli` to bridge the tokens with the command below:

```bash

> npx zksync-cli bridge deposit --token 0x_YOUR_TOKEN_ADDRESS --rpc=http://localhost:3050 --l1-rpc=http://localhost:8545

Reviewer: ```suggestion
npx zksync-cli bridge deposit --token <0x_YOUR_TOKEN_ADDRESS> \
--rpc=http://localhost:3050 \
--l1-rpc=http://localhost:8545
```

```

```shell
? Amount to deposit 5

> ? Private key of the sender 0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110 

Reviewer: ```suggestion
? Private key of the sender *** 
```

? Recipient address on L2 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049
```

To verify that this worked, let's check the new balance of our address on the L2 chain:

```bash

> zksync-cli wallet balance --address 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049 --rpc http://localhost:3050

Reviewer: ```suggestion
zksync-cli wallet balance \
--address 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049 \
--rpc http://localhost:3050
```

```

Note that we aren't checking for any specific token address here.

> The CLI will display the amount as ETH, but in reality, it's our bridged ERC20 tokens.

Reviewer: I'm guessing this is a thing we need to fix with the zksync-cli?

Author: I think we could definitely make an improvement here

And that's it! You are now running a local ZK chain that uses a custom ERC20 base token.
You can follow the steps in the previous section to deploy a contract to the new chain.

## Bridging ETH to your chain

> If your ERC20 token bridged to the L2 "looks" like ETH, then what happens if you bridge ETH to your custom chain?

Reviewer: Maybe
```suggestion
If your ERC20 token bridged to the L2 "acts" like ETH, then what happens if you bridge ETH to your custom chain?
```

Not sure whichever is better, using "looks" makes me thing that this is still in relation to the earlier comment about zksync-cli display it as ETH.

The answer is that if you try to bridge regular ETH to this chain, it will just use a different token contract address.

You can try this out by depositing ETH from the L1 to this new chain:

```bash

> npx zksync-cli bridge deposit --rpc=http://localhost:3050 --l1-rpc=http://localhost:8545

Reviewer: ```suggestion
npx zksync-cli bridge deposit \
--rpc=http://localhost:3050 \
--l1-rpc=http://localhost:8545
```

```

```shell
? Amount to deposit 20

> ? Private key of the sender 0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110 

Reviewer: ```suggestion
? Private key of the sender *** 
```

Author: see related comment above

? Recipient address on L2 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049
```

> If you run the previous command to check the base token balance again, note how the balance is still `5`.
> This is because you have to look at the L2 token balance for ETH.
> To find the L2 token address for ETH, you can use the `l2TokenAddress` method available through `zksync-ethers`.

Reviewer: ```suggestion
If you run the previous command to check the base token balance again, note how the balance is still `5`. Because ETH is no longer the base token, it has a different token address on L2 you will need to reference.
To find the L2 token address for ETH, you can use the `l2TokenAddress` method available through `zksync-ethers`.
```

To try this out, open the `zk-chain-test` folder from the previous section,
and run the commands below to create a new script file:

```bash
mkdir scripts
touch scripts/checkBalance.ts
```

> Next, copy and paste the script below to into the `checkBalance.ts` file:

Reviewer: ```suggestion
Next, copy and paste the script below into the `checkBalance.ts` file:
```

```ts
import { ETH_ADDRESS_IN_CONTRACTS } from "zksync-ethers/build/utils.js";
import { getWallet } from "../deploy/utils";

async function main(){
    const wallet = getWallet();
    const l2ETHAddress = await wallet.l2TokenAddress(ETH_ADDRESS_IN_CONTRACTS);
    console.log("L2 ETH Address:", l2ETHAddress);

    const balance = await wallet.getBalance(l2ETHAddress);
    console.log("L2 ETH Balance 🎉:", balance);
}

main();
```

Run the script with the `hardhat run` command:

```bash
npx hardhat run scripts/checkBalance.ts
```

You should see the output below showing the same amount of ETH you bridged:

```shell
L2 ETH Address: 0x_YOUR_ETH_TOKEN_ADDRESS
L2 ETH Balance 🎉: 20000000000000000000n
```

## Switching between chains and shutting down the ecosystem

> You can switch in between different chains without losing any state by shutting down the chain server and running the command below:

Reviewer: Does this mean the command below will shut down the chain server, or is there a specific way to do that and the command below is just for switching chains?

```ts
import { buildModule } from "@nomicfoundation/hardhat-ignition/modules";

const CustomBaseTokenModule = buildModule("CustomBaseTokenModule", (m) => {
  const tokenName = m.getParameter("name", "ZK Base Token");
  const tokenSymbol = m.getParameter("symbol", "ZKBT");

  const baseToken = m.contract("CustomBaseToken", [tokenName, tokenSymbol]);

  return { baseToken };
});

export default CustomBaseTokenModule;
```

To deploy the token contract, run:

```bash
npx hardhat ignition deploy ./ignition/modules/CustomBaseToken.ts --network localRethNode
```

Select `y` to confirm to deploy the contract to the `localRethNode`.
After deploying, the token contract address should be logged in your console.

Let's verify that the tokens were minted to the deployer address using the Foundry `cast` CLI:

```bash
cast balance --erc20 <0xYOUR_TOKEN_ADDRESS> 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049 \
--rpc-url http://localhost:8545
```

Now that you have some tokens, the next step is to send some to each of the governor's addresses for the ecosystem and chain on the L1.
This will allow you to register and deploy your chain, and deploy the paymaster.

> You can find the address for the ecosystem governor in `my_elastic_chain/configs/wallets.yaml` under `governor`.

Reviewer: ```suggestion
You can find the address for the ecosystem governor in `<my_elastic_chain>/configs/wallets.yaml` under `governor`.
```

> The chain governor address can be found in `my_elastic_chain/chains/custom_zk_chain/configs/wallets.yaml` under `governor`.

Reviewer: Same for here: <my_elastic_chain>

Run the command below to use `cast` to transfer some of your ERC20 tokens to the governor's address on the L1.

```bash
cast send <0xYOUR_TOKEN_ADDRESS> "transfer(address,uint256)" \
<0x_YOUR_GOVERNOR_ADDRESS> 1000000000000000000 \
--private-key 0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110 \
--rpc-url http://localhost:8545 \
--gas-price 30000000000
```

To verify that this worked, check the balance of the governor's address:

```bash
cast balance --erc20 <0xYOUR_TOKEN_ADDRESS> \
<0x_YOUR_GOVERNOR_ADDRESS> \
--rpc-url http://localhost:8545
```

## Creating a new chain

Now that your ERC20 token is deployed and the governor addresses are funded with that token, you can create a new chain.

First, shut down the node server running for `zk_chain_1`.

> Move back into the `my_elastic_chain` ecosystem folder and run the `chain create` command:

Reviewer: ```suggestion
Move back into your elastic chain ecosystem folder and run the `chain create` command:
```

    const balance = await wallet.getBalance(l2ETHAddress);
    console.log("L2 ETH Balance 🎉:", balance);
}

main();
```

Run the script with the `hardhat run` command:

```bash
npx hardhat run scripts/checkBalance.ts
```

You should see the output below showing the same amount of ETH you bridged:

```shell
L2 ETH Address: <0x_YOUR_ETH_TOKEN_ADDRESS>
L2 ETH Balance 🎉: 20000000000000000000n
```

## Switching between chains and shutting down the ecosystem

You can switch in between different chains without losing any state by shutting down the chain server and then running the command below:

```bash
zk_inception ecosystem change-default-chain
```

Now when you restart the server with `zk_inception server`, the new default chain's node will start again with the same state.

To fully shut down the ecosystem and erase all of the data and state,
you can shut down the Docker containers from the ecosystem folder using the command:

```bash
docker-compose down
```

To restart the ecosystem and run your custom ZK chain again, follow the steps below:

1. In the ecosystem folder, run `zk_inception containers`.
1. Redeploy your ERC20 contract to the L1.

> 1. Update the base token address in `my_elastic_chain/chains/custom_zk_chain/configs/contracts.yaml` under `l1.base_token_addr` and in
>   `my_elastic_chain/chains/custom_zk_chain/ZkStack.yaml` under `base_token.address`.

Reviewer: ```suggestion
1. Update the base token address in `<my_elastic_chain>/chains/custom_zk_chain/configs/contracts.yaml` under `l1.base_token_addr` and in
  `<my_elastic_chain>/chains/custom_zk_chain/ZkStack.yaml` under `base_token.address`.
```

Author: I think this one if ok since I have that folder name in the setup, so the user doesn't need to change this

First, shut down the node server running for `zk_chain_1`.

Move back into the `my_elastic_chain` ecosystem folder and run the `chain create` command:

```bash
zk_inception chain create
```

This time, use the answers below:

```bash
$ zk_inception chain create

┌   ZKsync toolbox
│
◇  What do you want to name the chain?
│  custom_zk_chain
│
◇  What's the chain id?
│  272
│
◇  Select how do you want to create the wallet
│  Localhost
│
◇  Select the prover mode
│  NoProofs
│
◇  Select the commit data generator mode
│  Rollup
│
◇  Select the base token to use
│  Custom
│
◇  What is the token address?
│  <0x_YOUR_TOKEN_ADDRESS>
│
◇  What is the base token price nominator?
│  1
│
◇  What is the base token price denominator?
│  1
│
◇  Set this chain as default?
│  Yes
```

The base token nominator and denominator are used to define the relation of the base token price with ETH.
For example, if we set the nominator to 20 and the denominator to 1, together the relation is 20/1,
this would mean that 20 tokens would be given the equivalent value as 1 ETH for gas.
For testing purposes, we'll just use a 1:1 ratio.

Next, initialize the chain in the ecosystem with:

```bash

> zk_inception chain init

Reviewer 2: it won't work our of the box, you have to top up configs/wallets.yaml/governor and chains/<name>/configs/wallets.yaml/governor

Author: it seems to work just funding the address in `configs/wallets.yaml/governor`, which I have as a step already

Author: Ahh I see that funding the chain governor allows you to deploy the paymaster in the setup. I will update this.

```

> This time select *no* for the option to deploy a paymaster contract.

Reviewer 2: why?

Author: I'll add a note about this, but from it seems that this involves funding another address with the erc20 base token, which seemed didn't seem like a necessary step for this tutorial

Author: I realized that was just the other governor's address - I removed this

---
title: Customizing your ZK chain
description: Create a ZK chain with a custom ERC20 base token.

---

One way you can customize your ZK chain is by changing the base token used for gas from ETH to an ERC20 token.
Let's try customizing a new chain by using a custom ERC20 base token.

With the initial release of ZK stack, custom ERC20 base tokens must be added to an allowlist for use as a base token for a chain.
The allowed addresses are stored and checked in the `BridgeHub` contract on the L1.

> In the future it will be possible to add a new token in a permissionless process.

Author: I'm not exactly sure if there will still be an allowlist then, but I will reword this to mean that you can deploy a chain with a custom base token in a permissionless process

> For now, you have the ability to add any tokens to the allowlist in your local ecosystem.

Author: it sort of expands on it - the sentence above says there is a limitation that will be changed in the future, and this says you can still try it out locally (since you run the ecosystem)

The overall flow for setting up a local chain with a custom ERC20 base token looks like this:

1. Deploy an ERC20 token to the L1.
1. Create your new chain that uses the ERC20 token and set it as the default.
1. Send ERC20 tokens to the ecosystem and chain governor addresses on the L1.
1. Initialize the new chain in the ecosystem.

> 1. Bridge tokens from the L1 to your new chain.

Reviewer: Where in this process is the ERC20 token added in the allowlist on the BridgeHub contract? It looks like from the rest of the guide that we ourselves aren't adding it to to the allowlist, so is it a part of the initialization of a ZK Chain?

Author: you don't have to do it directly, it happens in the background when you initialize the chain. The part at the top is there mainly to say that although we can do all of this in a tutorial, it's not a permissionless process to do it on testnet or mainnet

Reviewer: OH now I get it. It feels like there's a lot being indirectly addressed in the description above. So in production, you have to go through a process to get a token added to an allowlist to be used as a base token, but with the local installation we automate this allowlisting process for development ease, is that correct?

```bash
npx hardhat init
```

Select the option **Create a Typescript project** and accept the default options for everything else.

Run the command below to install the necessary dependencies:

::code-group

```bash [npm]
npm i -D typescript ts-node @openzeppelin/contracts @nomicfoundation/hardhat-ethers @typechain/ethers-v6 @typechain/hardhat typechain ethers dotenv
```

```bash [yarn]
yarn add -D typescript ts-node @openzeppelin/contracts @nomicfoundation/hardhat-ethers @typechain/ethers-v6 @typechain/hardhat typechain ethers dotenv
```

::

### Configuring Hardhat

Once installed, replace your existing config in `hardhat.config.ts` with the config below:

```ts
import { HardhatUserConfig } from "hardhat/config";
import "@nomicfoundation/hardhat-toolbox";

import dotenv from "dotenv";
dotenv.config();

const config: HardhatUserConfig = {
  solidity: "0.8.20",
  networks: {
        localRethNode: {
          url: "http://localhost:8545",
          accounts: [process.env.WALLET_PRIVATE_KEY as any],
        },
      },
};

export default config;
```

Now hardhat is configured to connect to our local reth node (the L1) running at local port `8545`.

> For the inital release of the ZK stack, we have to deploy the token to the L1 and bridge to our chain later.

Reviewer: ```suggestion
For the inital release of the ZK stack, we have to deploy the token to the L1 and bridge to our ZK chain later.
```

        _mint(msg.sender, 100 * 10 ** decimals());
    }

    function mint(address to, uint256 amount) public onlyOwner {
        _mint(to, amount);
    }
}
```

Next, let's update the ignition module to deploy the ERC20 contract.
Rename the module file with the command below:

```bash
mv ignition/modules/Lock.ts ignition/modules/CustomBaseToken.ts
```

Then, replace the module file with the code below:

```ts
import { buildModule } from "@nomicfoundation/hardhat-ignition/modules";

const CustomBaseTokenModule = buildModule("CustomBaseTokenModule", (m) => {
  const tokenName = m.getParameter("name", "ZK Base Token");
  const tokenSymbol = m.getParameter("symbol", "ZKBT");

  const baseToken = m.contract("CustomBaseToken", [tokenName, tokenSymbol]);

  return { baseToken };
});

export default CustomBaseTokenModule;
```

To run the module and deploy the token contract, run:

```bash
npx hardhat ignition deploy ./ignition/modules/CustomBaseToken.ts --network localRethNode
```

Select `y` to confirm to deploy the contract to the `localRethNode`.
After deploying, the token contract address should be logged in your console.

> Let's verify that the tokens were minted to the deployer address using the Foundry `cast` CLI:

Reviewer: When were tokens minted for the deployer address? Is that part of the ignition module?

Author: yes, the ignition module is deploying the token

Reviewer: ah, I glossed over the token contract

```bash
cast send <0xYOUR_TOKEN_ADDRESS> "transfer(address,uint256)" \
<0x_YOUR_GOVERNOR_ADDRESS> 1000000000000000000 \
--private-key 0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110 \
--rpc-url http://localhost:8545 \
--gas-price 30000000000
```

To verify that this worked, check the balance of both governor addresses:

```bash
cast balance --erc20 <0xYOUR_TOKEN_ADDRESS> \
<0x_YOUR_GOVERNOR_ADDRESS> \
--rpc-url http://localhost:8545
```

### Initializing the chain

::callout{icon="i-heroicons-exclamation-triangle" color="amber"}
Make sure the server for `zk_chain_1` that you started in the previous section is shut down.
::

Next, initialize the chain in the ecosystem with the command below, and select the default options for the prompts.

```bash
zk_inception chain init
```

Now that the chain is initialized, you can start the chain server:

```bash
zk_inception server
```

## Bridging the base token to your chain

The last step is to bridge some ERC20 tokens from the L1 to the new chain.
Base tokens can not be minted on the L2 without being backed by the corresponding L1 amount. In the future, there will be more flexibility for this.

> Open a new terminal and use the `zksync-cli` to bridge the tokens with the command below:

Reviewer: ```suggestion
Open a new terminal and use ZKsync CLI to bridge the tokens with the command below:
```

And that's it! You are now running a local ZK chain that uses a custom ERC20 base token.
You can follow the steps in the previous section to deploy a contract to the new chain.

## Bridging ETH to your chain

If your ERC20 token bridged to the L2 "acts" like ETH, then what happens if you bridge ETH to your custom chain?

The answer is that if you try to bridge regular ETH to this chain, it will just use a different token contract address.

You can try this out by depositing ETH from the L1 to this new chain:

```bash
npx zksync-cli bridge deposit \
--rpc=http://localhost:3050 \
--l1-rpc=http://localhost:8545
```

```shell
? Amount to deposit 20
? Private key of the sender 0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110
? Recipient address on L2 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049
```

If you run the previous command to check the base token balance again, note how the balance is still `5`.
Because ETH is no longer the base token, it has a different token address on L2 you will need to reference.
To find the L2 token address for ETH, you can use the `l2TokenAddress` method available through `zksync-ethers`.

> To try this out, open the `zk-chain-test` folder from the previous section,

Reviewer: ```suggestion
To try this out, open your hardhat project `zk-chain-test` from the previous section,
```

and run the commands below to create a new script file:

```bash
mkdir scripts
touch scripts/checkBalance.ts
```

Next, copy and paste the script below into the `checkBalance.ts` file:

```ts
import { ETH_ADDRESS_IN_CONTRACTS } from "zksync-ethers/build/utils.js";
import { getWallet } from "../deploy/utils";

async function main(){
    const wallet = getWallet();
    const l2ETHAddress = await wallet.l2TokenAddress(ETH_ADDRESS_IN_CONTRACTS);
    console.log("L2 ETH Address:", l2ETHAddress);

    const balance = await wallet.getBalance(l2ETHAddress);
    console.log("L2 ETH Balance 🎉:", balance);
}

main();
```

Run the script with the `hardhat run` command:

```bash
npx hardhat run scripts/checkBalance.ts
```

You should see the output below showing the same amount of ETH you bridged:

```shell
L2 ETH Address: <0x_YOUR_ETH_TOKEN_ADDRESS>
L2 ETH Balance 🎉: 20000000000000000000n
```

## Switching between chains and shutting down the ecosystem

> You can switch in between different chains without losing any state by shutting down the chain server and then running the command below:

Reviewer: How do we shut down the chain server?

```bash
zk_inception ecosystem change-default-chain
```

> Now when you restart the server with `zk_inception server`, the new default chain's node will start again with the same state.

Reviewer: ```suggestion
Now when you start the server with `zk_inception server`, the new default chain's node will start with the saved state.
```

To fully shut down the ecosystem and erase all of the data and state,
you can shut down the Docker containers from the ecosystem folder using the command:

```bash
docker-compose down
```

> To restart the ecosystem and run your custom ZK chain again, follow the steps below:

Reviewer: ```suggestion
To restart the docker containers for your ecosystem and run your custom ZK chain again, follow the steps below:
```

To find the L2 token address for ETH, you can use the `l2TokenAddress` method available through `zksync-ethers`.

To try this out, open your hardhat project `zk-chain-test` from the previous section,
and run the commands below to create a new script file:

```bash
mkdir scripts
touch scripts/checkBalance.ts
```

Next, copy and paste the script below into the `checkBalance.ts` file:

```ts
import { ETH_ADDRESS_IN_CONTRACTS } from "zksync-ethers/build/utils.js";
import { getWallet } from "../deploy/utils";

async function main(){
    const wallet = getWallet();
    const l2ETHAddress = await wallet.l2TokenAddress(ETH_ADDRESS_IN_CONTRACTS);
    console.log("L2 ETH Address:", l2ETHAddress);

    const balance = await wallet.getBalance(l2ETHAddress);
    console.log("L2 ETH Balance 🎉:", balance);
}

main();
```

Run the script with the `hardhat run` command:

```bash
npx hardhat run scripts/checkBalance.ts
```

You should see the output below showing the same amount of ETH you bridged:

```shell
L2 ETH Address: <0x_YOUR_ETH_TOKEN_ADDRESS>
L2 ETH Balance 🎉: 20000000000000000000n
```

## Switching between chains and shutting down the ecosystem

> You can switch in between different chains without losing any state by terminating the chain server process and then running the command below:

Reviewer: Is there a command to terminate the chain server process or do I stop a docker container somewhere?

Reviewer 3: I was using this from Danil's suggestion: `pkill -9 zksync_server`

Author: I use control C on my mac but I'm guessing that doesn't work on every machine. It's not a docker container it's just a process running in a terminal

Reviewer: Ah, didn't know it was an actively running process in a terminal window. 👍🏻
